src/vanilla/utils/selectAtom.ts

```diff
@@ -15,39 +15,51 @@ const memo3 = <T>(
   return getCached(create, cache3, dep3)
 }
 
+type PromiseOrValue<T> = T | Promise<T>
+
 export function selectAtom<Value, Slice>(
   anAtom: Atom<Value>,
-  selector: (v: Awaited<Value>, prevSlice?: Slice) => Slice,
+  selector: (v: Value, prevSlice?: Slice) => Slice,
   equalityFn?: (a: Slice, b: Slice) => boolean,
-): Atom<Value extends Promise<unknown> ? Promise<Slice> : Slice>
+): Atom<Slice>
+
+export function selectAtom<Value, Slice extends Promise<unknown>>(
+  anAtom: Atom<Value>,
+  selector: (v: Value, prevSlice?: Slice) => Slice,
+  equalityFn?: (a: Slice, b: Slice) => Promise<boolean>,
+): Atom<Promise<Slice>>
 
 export function selectAtom<Value, Slice>(
   anAtom: Atom<Value>,
-  selector: (v: Awaited<Value>, prevSlice?: Slice) => Slice,
-  equalityFn: (a: Slice, b: Slice) => boolean = Object.is,
+  selector: (v: Value, prevSlice?: PromiseOrValue<Slice>) => Slice,
+  equalityFn: (
+    prevSlice: PromiseOrValue<Slice>,
+    slice: PromiseOrValue<Slice>,
+  ) => PromiseOrValue<boolean> = Object.is,
 ) {
   return memo3(
     () => {
       const EMPTY = Symbol()
       const selectValue = ([value, prevSlice]: readonly [
-        Awaited<Value>,
-        Slice | typeof EMPTY,
+        Value,
+        PromiseOrValue<Slice> | typeof EMPTY,
       ]) => {
         if (prevSlice === EMPTY) {
           return selector(value)
         }
         const slice = selector(value, prevSlice)
-        return equalityFn(prevSlice, slice) ? prevSlice : slice
+        const areEqual = equalityFn(prevSlice, slice)
+        if (areEqual instanceof Promise) {
+          return areEqual.then((areEqual) => (areEqual ? prevSlice : slice))
+        }
```

Comment:
Are we still resolving promises for equalityFn? I think it's one of complexity we want to avoid. (And, I think we'd generally recommend to use `unwrap` for promise-usage with `selectoAtom`.)

Reply:
> Are we still resolving promises for equalityFn? I think it's one of complexity we want to avoid. (And, I think we'd generally recommend to use `unwrap` for promise-usage with `selectoAtom`.)

Well, the main difference is that for equalityFn to work, the user would need to await the prevSlice if it were a promise.

Reply:
As we are making breaking changes anyway, I hope to drop that capability too and recommend `unwrap`ping promises in advance.